Auth: conflicting users validation improvements

eleijonmarck opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**
The validation command was giving cryptic errors when submitting different formats given in the edited file. This gives more context for failure of parsing for valid input from the user.

This adds additional validation for the validation of the file input to the validation function of conflicting users.

- it checks for any malformed input of the lines. we expect either a conflict block, an addition or a removal
- adds check to ensure that we only have one user to keep for each block
- adds an example terraform configuration of users found as example in the blogpost showcasing the different types of conflicts

**why**
Since we allow users to edit a file we need more careful validation of the input that we consider to be valid from the generated file. We also want to make it easier for customers to see what we mean by conflict by introducing an example file

---
Screenshots of the validation improvements

<img width="1136" alt="image" src="https://user-images.githubusercontent.com/7727602/199720127-7029921e-2bd3-4707-8b9c-032ec496da30.png">
<img width="1171" alt="image" src="https://user-images.githubusercontent.com/7727602/199720036-ee00f21a-6b92-4383-a16c-44b7fc4be789.png">

eleijonmarck committed:
WIP

eleijonmarck committed:
add: better validation of conflict file

eleijonmarck committed:
Merge branch 'main' into eleijonmarck/auth/conflicts-validation-improvement

eleijonmarck committed:
add: better description of validation and ingest command

eleijonmarck committed:
add: check for at least one user to delete

eleijonmarck committed:
add: example in terraform to setup for conflicts

IevaVasiljeva commented on pkg/cmd/grafana-cli/commands/conflict_user_command.go:
```diff
@@ -269,6 +288,18 @@ func getValidConflictUsers(r *ConflictResolver, b []byte) error {
 	}
+	for _, count := range counterKeepUsersForBlock {
+		// check if we only have one addition for each block
+		if count != 1 {
+			return fmt.Errorf("invalid number of users to keep, expected 1, got %d", count)
+		}
```
thought: would it be worth printing which block has invalid number of users added/removed here and below as well?

eleijonmarck replied:
great thought! 💭

eleijonmarck committed:
Update pkg/cmd/grafana-cli/commands/conflict_user_command.go

Co-authored-by: Ieva <ieva.vasiljeva@grafana.com>

eleijonmarck committed:
Add: print of conflict block for error

- adds conflict block to error output for validation of the file to
  easier diagnose in the file

eleijonmarck committed:
Merge branch 'main' into eleijonmarck/auth/conflicts-validation-improvement

eleijonmarck committed:
fix: formatting of errors

eleijonmarck committed:
fix: info strings improvements

eleijonmarck committed:
add: default 0 to blocks to check for users

eleijonmarck committed:
fixed: tests

eleijonmarck committed:
Merge branch 'main' into eleijonmarck/auth/conflicts-validation-improvement

eleijonmarck committed:
test integration

eleijonmarck committed:
fix strings fmt

eleijonmarck commented:
found a bug that needs to be fixed before merging this! 🙏

eleijonmarck committed:
set store in resolver
